src/python_pachyderm/mixin/pfs.py

```diff
@@ -1,60 +1,31 @@
 import io
+import os
 import re
 import itertools
 import tarfile
 from contextlib import contextmanager
 from typing import Iterator, Union, List, BinaryIO
 
+import grpc
+from betterproto import BytesValue
+
 from python_pachyderm.pfs import commit_from, uuid_re, SubcommitType
 from python_pachyderm.service import pfs_proto, Service
 from google.protobuf import empty_pb2, wrappers_pb2
 
-
 BUFFER_SIZE = 19 * 1024 * 1024
 
 
-class FileTarstream:
-    """Implements a file-like interface over a GRPC byte stream,
-    so we can use tarfile to decode the file contents.
-    """
-
-    def __init__(self, res):
-        self.res = res
-        self.buf = []
-
-    def __next__(self):
-        return next(self.res).value
-
-    def close(self):
-        self.res.cancel()
-
-    def read(self, size=-1):
-        if self.res.cancelled():
-            return b""
-
-        buf = []
-        remaining = size if size >= 0 else 2 ** 32
-
-        if self.buf:
-            buf.append(self.buf[:remaining])
-            self.buf = self.buf[remaining:]
-            remaining -= len(buf[-1])
-
-        try:
-            while remaining > 0:
-                b = next(self)
-
-                if len(b) > remaining:
-                    buf.append(b[:remaining])
-                    self.buf = b[remaining:]
-                else:
-                    buf.append(b)
-
-                remaining -= len(buf[-1])
-        except StopIteration:
-            pass
-
-        return b"".join(buf)
+class PfsTarFile(tarfile.TarFile):
+    def __iter__(self):
```

Comment:
Is this called during `extractall()`?

Reply:
Yeah, the inner working of `tarfile.TarFile` is kind of strange, it iterates over itself to get the individual member of the TAR. This seemed like the easiest way to insert custom code into that

Reply:
Solid